lib/jsonrpc_client.ml

```diff
@@ -0,0 +1,83 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+(* JSON-RPC Client *)
+
+module D = Debug.Make(struct let name = "jsonrpc_client" end)
+open D
+
+let input_json_object fin =
+	let buf = Buffer.create 1024 in
+	let brace_cnt = ref 0 in
+	let in_string = ref false in
+	let last_char () = Buffer.nth buf (Buffer.length buf - 1) in
+	let rec get () =
+		let c = input_char fin in
+		begin
+			match c with
+			| '{' when not !in_string -> brace_cnt := !brace_cnt + 1
+			| '}' when not !in_string -> brace_cnt := !brace_cnt - 1
+			| '"' when !in_string && (last_char () <> '\\') -> in_string := false
+			| '"' when not !in_string -> in_string := true
+			| _ -> ()
+		end;
+		Buffer.add_char buf c;
+		if !brace_cnt > 0 then
+			get ()
+	in
+	get ();
+	Buffer.contents buf
+
+let received = ref None
+let c = Condition.create ()
+let m = Mutex.create ()
+
+let rec receive fin () =
+	try
+		let obj = input_json_object fin in
+		debug "Received: %s" obj;
+		Mutex.lock m;
+		received := Some (Jsonrpc.response_of_string obj);
+		Condition.signal c;
+		Mutex.unlock m;
+		receive fin ()
+	with _ ->
+		(* end of file... stop thread *)
+		()
+
+let with_connection socket f =
+	let fin, fout = Unix.open_connection socket in
+	debug "Connected.";
+	let (_ : Thread.t) = Thread.create (receive fin) () in
+	let result = f fin fout in
+	Unix.shutdown_connection fin;
```

Comment:
It appears that I just need to close fin, after which the socket gets closed down (from https://ocaml.github.io/ocamlunix/ocamlunix.html#sec134):

> The client can shutdown the connection abruptly by closing either of the channels (this will close the socket) or more “cleanly” by calling shutdown_connection.